{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# How to use EBI Metagenomics API\n",
    "\n",
    "The EMG REST API https://www.ebi.ac.uk/metagenomics/api/ provides an easy-to-use set of top level resources, such as studies, samples, runs, experiment-types, biomes and annotations, that let user access metagenomics data in simple JSON format (JSON is a syntax for storing and exchanging data). Retrieving the data is as simple as sending a HTTP request. Response return JSON object formatted data structure that contains the resource type, associated object identifier (id) with attributes. Where appropriate, relationships and links are provided to other resources.\n",
    "\n",
    "We have utilised an interactive documentation framework (Swagger UI) to visualise and simplify interaction with the API’s resources via an HTML interface. Detailed explanations of the purpose of all resources, along with many examples, are provided to guide end-users. Documentation on how to use the endpoints is available at https://www.ebi.ac.uk/metagenomics/api/docs/."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Browse API"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 1\n",
    "\n",
    "Find metagenomic studies\n",
    "\n",
    "Answer: https://www.ebi.ac.uk/metagenomics/api/v0.2/studies?experiment_type=metagenomic"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 2\n",
    "\n",
    "Find oceanic metagenomic samples taken from latitude >= 70° (N)\n",
    "\n",
    "Answer: https://www.ebi.ac.uk/metagenomics/api/v0.2/samples?experiment_type=metagenomic&lineage=root%3AEnvironmental%3AAquatic%3AMarine%3AOceanic&latitude_gte=70"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write scripts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import Python modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from pandas import DataFrame\n",
    "\n",
    "try:\n",
    "    from urllib import urlencode\n",
    "except ImportError:\n",
    "    from urllib.parse import urlencode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from jsonapi_client import Session, Filter\n",
    "\n",
    "API_BASE = 'https://www.ebi.ac.uk/metagenomics/api/v0.2/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get study and list samples with biome\n",
    "\n",
    "Get study: https://www.ebi.ac.uk/metagenomics/api/v0.2/studies/ERP005831\n",
    "\n",
    "List samples: https://www.ebi.ac.uk/metagenomics/api/v0.2/studies/ERP005831/samples\n",
    "\n",
    "Change study accession to ERP002497"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample name</th>\n",
       "      <th>biome</th>\n",
       "      <th>lineage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERS456668</th>\n",
       "      <td>Tocil Lake surface sediment (12C)</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>root:Environmental:Aquatic:Freshwater:Lentic:S...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERS456669</th>\n",
       "      <td>Hunts Mill field soil (12C)</td>\n",
       "      <td>Agricultural</td>\n",
       "      <td>root:Environmental:Terrestrial:Soil:Loam:Agric...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 sample name         biome  \\\n",
       "accession                                                    \n",
       "ERS456668  Tocil Lake surface sediment (12C)      Sediment   \n",
       "ERS456669        Hunts Mill field soil (12C)  Agricultural   \n",
       "\n",
       "                                                     lineage  \n",
       "accession                                                     \n",
       "ERS456668  root:Environmental:Aquatic:Freshwater:Lentic:S...  \n",
       "ERS456669  root:Environmental:Terrestrial:Soil:Loam:Agric...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = DataFrame(columns=('sample name', 'biome', 'lineage'))\n",
    "df.index.name = 'accession'\n",
    "\n",
    "with Session(API_BASE) as s:\n",
    "    study = s.get('studies', 'ERP005831').resource\n",
    "    for sample in study.samples:\n",
    "        biome = sample.biome\n",
    "        df.loc[sample.accession] = [sample.sample_name, biome.biome_name, biome.lineage]\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### List runs\n",
    "\n",
    "Get sample: https://www.ebi.ac.uk/metagenomics/api/v0.2/samples/ERS667565\n",
    "\n",
    "List runs: https://www.ebi.ac.uk/metagenomics/api/v0.2/samples/ERS667565/runs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>instrument platform</th>\n",
       "      <th>instrument model</th>\n",
       "      <th>analysis pipeline</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERR867951</th>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR867950</th>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR771104</th>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2.0, 4.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          instrument platform instrument model analysis pipeline\n",
       "accession                                                       \n",
       "ERR867951            ILLUMINA   Illumina MiSeq               4.0\n",
       "ERR867950            ILLUMINA   Illumina MiSeq               4.0\n",
       "ERR771104            ILLUMINA   Illumina MiSeq          2.0, 4.0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = DataFrame(columns=('instrument platform', 'instrument model', 'analysis pipeline'))\n",
    "df.index.name = 'accession'\n",
    "\n",
    "with Session(API_BASE) as s:\n",
    "    sample = s.get('samples', 'ERS667565').resource\n",
    "    for run in sample.runs:\n",
    "        df.loc[run.accession] = [\n",
    "            run.instrument_platform, run.instrument_model,\n",
    "            \", \".join([p.release_version for p in run.pipelines])\n",
    "        ]\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### List sample metadata\n",
    "\n",
    "Get sample: https://www.ebi.ac.uk/metagenomics/api/v0.2/samples/ERS488919\n",
    "\n",
    "List sample metadata: https://www.ebi.ac.uk/metagenomics/api/v0.2/samples/ERS488919/metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TARA_20100318T1133Z_039_EVENT_PUMP_P_D_(25 m)_BACT_NUC-DNA(100L)_W1.6-20_TARA_B100000105 ERS488919\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>metadata key</th>\n",
       "      <th>value</th>\n",
       "      <th>unit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>temperature</td>\n",
       "      <td>26.812225</td>\n",
       "      <td>°C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>project name</td>\n",
       "      <td>Tara Oceans expedition (2009-2013)</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>geographic location (depth)</td>\n",
       "      <td>25</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>environmental package</td>\n",
       "      <td>water</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>instrument model</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENA checklist</td>\n",
       "      <td>ENA TARA (ERC000030)</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>latitude end</td>\n",
       "      <td>18.5679</td>\n",
       "      <td>DD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>longitude end</td>\n",
       "      <td>66.4581</td>\n",
       "      <td>DD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>marine region</td>\n",
       "      <td>n/a</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>protocol label</td>\n",
       "      <td>BACT_NUC-DNA(100L)_W1.6-20</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>sampling campaign</td>\n",
       "      <td>TARA_20100309Z</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>sampling platform</td>\n",
       "      <td>SV Tara</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chlorophyll sensor</td>\n",
       "      <td>0.17638</td>\n",
       "      <td>mg Chl/m3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>citation</td>\n",
       "      <td>tbd</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>event date/time end</td>\n",
       "      <td>2010-03-18T12:32</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>event date/time start</td>\n",
       "      <td>2010-03-18T11:33</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>event label</td>\n",
       "      <td>TARA_20100318T1133Z_039_EVENT_PUMP</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>further details</td>\n",
       "      <td>tbd</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>last update date</td>\n",
       "      <td>2014-05-01Z</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>nitrate sensor</td>\n",
       "      <td>-0.611888</td>\n",
       "      <td>µmol/L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>oxygen sensor</td>\n",
       "      <td>192.95875</td>\n",
       "      <td>µmol/Kg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>salinity sensor</td>\n",
       "      <td>36.332317</td>\n",
       "      <td>psu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>sample collection device</td>\n",
       "      <td>PUMP (High Volume Peristaltic Pump) with ECOTr...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>sample status</td>\n",
       "      <td>This version can be used to provide data disco...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>sampling station</td>\n",
       "      <td>TARA_039</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>size fraction lower threshold</td>\n",
       "      <td>1.6</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>size fraction upper threshold</td>\n",
       "      <td>20</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     metadata key  \\\n",
       "0                     temperature   \n",
       "1                    project name   \n",
       "2     geographic location (depth)   \n",
       "3           environmental package   \n",
       "4                instrument model   \n",
       "5                   ENA checklist   \n",
       "6                    latitude end   \n",
       "7                   longitude end   \n",
       "8                   marine region   \n",
       "9                  protocol label   \n",
       "10              sampling campaign   \n",
       "11              sampling platform   \n",
       "12             chlorophyll sensor   \n",
       "13                       citation   \n",
       "14            event date/time end   \n",
       "15          event date/time start   \n",
       "16                    event label   \n",
       "17                further details   \n",
       "18               last update date   \n",
       "19                 nitrate sensor   \n",
       "20                  oxygen sensor   \n",
       "21                salinity sensor   \n",
       "22       sample collection device   \n",
       "23                  sample status   \n",
       "24               sampling station   \n",
       "25  size fraction lower threshold   \n",
       "26  size fraction upper threshold   \n",
       "\n",
       "                                                value       unit  \n",
       "0                                           26.812225         °C  \n",
       "1                  Tara Oceans expedition (2009-2013)             \n",
       "2                                                  25          m  \n",
       "3                                               water             \n",
       "4                                 Illumina HiSeq 2000             \n",
       "5                                ENA TARA (ERC000030)             \n",
       "6                                             18.5679         DD  \n",
       "7                                             66.4581         DD  \n",
       "8                                                 n/a             \n",
       "9                          BACT_NUC-DNA(100L)_W1.6-20             \n",
       "10                                     TARA_20100309Z             \n",
       "11                                            SV Tara             \n",
       "12                                            0.17638  mg Chl/m3  \n",
       "13                                                tbd             \n",
       "14                                   2010-03-18T12:32             \n",
       "15                                   2010-03-18T11:33             \n",
       "16                 TARA_20100318T1133Z_039_EVENT_PUMP             \n",
       "17                                                tbd             \n",
       "18                                        2014-05-01Z             \n",
       "19                                          -0.611888     µmol/L  \n",
       "20                                          192.95875    µmol/Kg  \n",
       "21                                          36.332317        psu  \n",
       "22  PUMP (High Volume Peristaltic Pump) with ECOTr...             \n",
       "23  This version can be used to provide data disco...             \n",
       "24                                           TARA_039             \n",
       "25                                                1.6             \n",
       "26                                                 20             "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def format_unit(unit):\n",
    "    import html\n",
    "    return html.unescape(m.unit) if m.unit else \"\"\n",
    "\n",
    "df = DataFrame(columns=('metadata key', 'value', 'unit'))\n",
    "\n",
    "with Session(API_BASE) as s:\n",
    "    sample = s.get('samples', 'ERS488919').resource\n",
    "    print(sample.sample_name, sample.accession)\n",
    "            \n",
    "    for i, m in enumerate(sample.metadata):\n",
    "        df.loc[i] = [\n",
    "            m.var_name, m.var_value,\n",
    "            format_unit(m.unit)\n",
    "        ]\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### List functional annotations\n",
    "\n",
    "Gene Ontology (GO) terms derived from InterPro matches: https://www.ebi.ac.uk/metagenomics/api/v0.2/runs/ERR263024/pipelines/1.0/go-slim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>description</th>\n",
       "      <th>annotation counts</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO term</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GO:0000156</th>\n",
       "      <td>molecular_function</td>\n",
       "      <td>two-component response regulator activity</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0000160</th>\n",
       "      <td>biological_process</td>\n",
       "      <td>phosphorelay signal transduction system</td>\n",
       "      <td>50.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0000166</th>\n",
       "      <td>molecular_function</td>\n",
       "      <td>nucleotide binding</td>\n",
       "      <td>316.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0000746</th>\n",
       "      <td>biological_process</td>\n",
       "      <td>conjugation</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0000902</th>\n",
       "      <td>biological_process</td>\n",
       "      <td>cell morphogenesis</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0000988</th>\n",
       "      <td>molecular_function</td>\n",
       "      <td>protein binding transcription factor activity</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0001071</th>\n",
       "      <td>molecular_function</td>\n",
       "      <td>nucleic acid binding transcription factor acti...</td>\n",
       "      <td>47.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0002376</th>\n",
       "      <td>biological_process</td>\n",
       "      <td>immune system process</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0003676</th>\n",
       "      <td>molecular_function</td>\n",
       "      <td>nucleic acid binding</td>\n",
       "      <td>218.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0003774</th>\n",
       "      <td>molecular_function</td>\n",
       "      <td>motor activity</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0003924</th>\n",
       "      <td>molecular_function</td>\n",
       "      <td>GTPase activity</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0004386</th>\n",
       "      <td>molecular_function</td>\n",
       "      <td>helicase activity</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0004518</th>\n",
       "      <td>molecular_function</td>\n",
       "      <td>nuclease activity</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0004672</th>\n",
       "      <td>molecular_function</td>\n",
       "      <td>protein kinase activity</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0004812</th>\n",
       "      <td>molecular_function</td>\n",
       "      <td>aminoacyl-tRNA ligase activity</td>\n",
       "      <td>48.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0004872</th>\n",
       "      <td>molecular_function</td>\n",
       "      <td>receptor activity</td>\n",
       "      <td>78.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0005102</th>\n",
       "      <td>molecular_function</td>\n",
       "      <td>receptor binding</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0005215</th>\n",
       "      <td>molecular_function</td>\n",
       "      <td>transporter activity</td>\n",
       "      <td>150.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0005618</th>\n",
       "      <td>cellular_component</td>\n",
       "      <td>cell wall</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0005634</th>\n",
       "      <td>cellular_component</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0005667</th>\n",
       "      <td>cellular_component</td>\n",
       "      <td>transcription factor complex</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0005694</th>\n",
       "      <td>cellular_component</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0005727</th>\n",
       "      <td>cellular_component</td>\n",
       "      <td>extrachromosomal circular DNA</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0005737</th>\n",
       "      <td>cellular_component</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>83.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0005739</th>\n",
       "      <td>cellular_component</td>\n",
       "      <td>mitochondrion</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0005773</th>\n",
       "      <td>cellular_component</td>\n",
       "      <td>vacuole</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0005783</th>\n",
       "      <td>cellular_component</td>\n",
       "      <td>endoplasmic reticulum</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0005794</th>\n",
       "      <td>cellular_component</td>\n",
       "      <td>Golgi apparatus</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0005840</th>\n",
       "      <td>cellular_component</td>\n",
       "      <td>ribosome</td>\n",
       "      <td>41.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0005856</th>\n",
       "      <td>cellular_component</td>\n",
       "      <td>cytoskeleton</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0019867</th>\n",
       "      <td>cellular_component</td>\n",
       "      <td>outer membrane</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0022900</th>\n",
       "      <td>biological_process</td>\n",
       "      <td>electron transport chain</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0030031</th>\n",
       "      <td>biological_process</td>\n",
       "      <td>cell projection assembly</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0030234</th>\n",
       "      <td>molecular_function</td>\n",
       "      <td>enzyme regulator activity</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0031012</th>\n",
       "      <td>cellular_component</td>\n",
       "      <td>extracellular matrix</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0031982</th>\n",
       "      <td>cellular_component</td>\n",
       "      <td>vesicle</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0032196</th>\n",
       "      <td>biological_process</td>\n",
       "      <td>transposition</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0034061</th>\n",
       "      <td>molecular_function</td>\n",
       "      <td>DNA polymerase activity</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0034062</th>\n",
       "      <td>molecular_function</td>\n",
       "      <td>RNA polymerase activity</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0042575</th>\n",
       "      <td>cellular_component</td>\n",
       "      <td>DNA polymerase complex</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0042592</th>\n",
       "      <td>biological_process</td>\n",
       "      <td>homeostatic process</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0042597</th>\n",
       "      <td>cellular_component</td>\n",
       "      <td>periplasmic space</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0043190</th>\n",
       "      <td>cellular_component</td>\n",
       "      <td>ATP-binding cassette (ABC) transporter complex</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0043934</th>\n",
       "      <td>biological_process</td>\n",
       "      <td>sporulation</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0044281</th>\n",
       "      <td>biological_process</td>\n",
       "      <td>small molecule metabolic process</td>\n",
       "      <td>303.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0044403</th>\n",
       "      <td>biological_process</td>\n",
       "      <td>symbiosis, encompassing mutualism through para...</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0045182</th>\n",
       "      <td>molecular_function</td>\n",
       "      <td>translation regulator activity</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0045333</th>\n",
       "      <td>biological_process</td>\n",
       "      <td>cellular respiration</td>\n",
       "      <td>19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0046034</th>\n",
       "      <td>biological_process</td>\n",
       "      <td>ATP metabolic process</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0046039</th>\n",
       "      <td>biological_process</td>\n",
       "      <td>GTP metabolic process</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0046872</th>\n",
       "      <td>molecular_function</td>\n",
       "      <td>metal ion binding</td>\n",
       "      <td>132.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0048037</th>\n",
       "      <td>molecular_function</td>\n",
       "      <td>cofactor binding</td>\n",
       "      <td>183.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0048870</th>\n",
       "      <td>biological_process</td>\n",
       "      <td>cell motility</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0051082</th>\n",
       "      <td>molecular_function</td>\n",
       "      <td>unfolded protein binding</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0051186</th>\n",
       "      <td>biological_process</td>\n",
       "      <td>cofactor metabolic process</td>\n",
       "      <td>48.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0051301</th>\n",
       "      <td>biological_process</td>\n",
       "      <td>cell division</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0055114</th>\n",
       "      <td>biological_process</td>\n",
       "      <td>oxidation-reduction process</td>\n",
       "      <td>325.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0070469</th>\n",
       "      <td>cellular_component</td>\n",
       "      <td>respiratory chain</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0071554</th>\n",
       "      <td>biological_process</td>\n",
       "      <td>cell wall organization or biogenesis</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0071941</th>\n",
       "      <td>biological_process</td>\n",
       "      <td>nitrogen cycle metabolic process</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>116 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      category  \\\n",
       "GO term                          \n",
       "GO:0000156  molecular_function   \n",
       "GO:0000160  biological_process   \n",
       "GO:0000166  molecular_function   \n",
       "GO:0000746  biological_process   \n",
       "GO:0000902  biological_process   \n",
       "GO:0000988  molecular_function   \n",
       "GO:0001071  molecular_function   \n",
       "GO:0002376  biological_process   \n",
       "GO:0003676  molecular_function   \n",
       "GO:0003774  molecular_function   \n",
       "GO:0003924  molecular_function   \n",
       "GO:0004386  molecular_function   \n",
       "GO:0004518  molecular_function   \n",
       "GO:0004672  molecular_function   \n",
       "GO:0004812  molecular_function   \n",
       "GO:0004872  molecular_function   \n",
       "GO:0005102  molecular_function   \n",
       "GO:0005215  molecular_function   \n",
       "GO:0005618  cellular_component   \n",
       "GO:0005634  cellular_component   \n",
       "GO:0005667  cellular_component   \n",
       "GO:0005694  cellular_component   \n",
       "GO:0005727  cellular_component   \n",
       "GO:0005737  cellular_component   \n",
       "GO:0005739  cellular_component   \n",
       "GO:0005773  cellular_component   \n",
       "GO:0005783  cellular_component   \n",
       "GO:0005794  cellular_component   \n",
       "GO:0005840  cellular_component   \n",
       "GO:0005856  cellular_component   \n",
       "...                        ...   \n",
       "GO:0019867  cellular_component   \n",
       "GO:0022900  biological_process   \n",
       "GO:0030031  biological_process   \n",
       "GO:0030234  molecular_function   \n",
       "GO:0031012  cellular_component   \n",
       "GO:0031982  cellular_component   \n",
       "GO:0032196  biological_process   \n",
       "GO:0034061  molecular_function   \n",
       "GO:0034062  molecular_function   \n",
       "GO:0042575  cellular_component   \n",
       "GO:0042592  biological_process   \n",
       "GO:0042597  cellular_component   \n",
       "GO:0043190  cellular_component   \n",
       "GO:0043934  biological_process   \n",
       "GO:0044281  biological_process   \n",
       "GO:0044403  biological_process   \n",
       "GO:0045182  molecular_function   \n",
       "GO:0045333  biological_process   \n",
       "GO:0046034  biological_process   \n",
       "GO:0046039  biological_process   \n",
       "GO:0046872  molecular_function   \n",
       "GO:0048037  molecular_function   \n",
       "GO:0048870  biological_process   \n",
       "GO:0051082  molecular_function   \n",
       "GO:0051186  biological_process   \n",
       "GO:0051301  biological_process   \n",
       "GO:0055114  biological_process   \n",
       "GO:0070469  cellular_component   \n",
       "GO:0071554  biological_process   \n",
       "GO:0071941  biological_process   \n",
       "\n",
       "                                                  description  \\\n",
       "GO term                                                         \n",
       "GO:0000156          two-component response regulator activity   \n",
       "GO:0000160            phosphorelay signal transduction system   \n",
       "GO:0000166                                 nucleotide binding   \n",
       "GO:0000746                                        conjugation   \n",
       "GO:0000902                                 cell morphogenesis   \n",
       "GO:0000988      protein binding transcription factor activity   \n",
       "GO:0001071  nucleic acid binding transcription factor acti...   \n",
       "GO:0002376                              immune system process   \n",
       "GO:0003676                               nucleic acid binding   \n",
       "GO:0003774                                     motor activity   \n",
       "GO:0003924                                    GTPase activity   \n",
       "GO:0004386                                  helicase activity   \n",
       "GO:0004518                                  nuclease activity   \n",
       "GO:0004672                            protein kinase activity   \n",
       "GO:0004812                     aminoacyl-tRNA ligase activity   \n",
       "GO:0004872                                  receptor activity   \n",
       "GO:0005102                                   receptor binding   \n",
       "GO:0005215                               transporter activity   \n",
       "GO:0005618                                          cell wall   \n",
       "GO:0005634                                            nucleus   \n",
       "GO:0005667                       transcription factor complex   \n",
       "GO:0005694                                         chromosome   \n",
       "GO:0005727                      extrachromosomal circular DNA   \n",
       "GO:0005737                                          cytoplasm   \n",
       "GO:0005739                                      mitochondrion   \n",
       "GO:0005773                                            vacuole   \n",
       "GO:0005783                              endoplasmic reticulum   \n",
       "GO:0005794                                    Golgi apparatus   \n",
       "GO:0005840                                           ribosome   \n",
       "GO:0005856                                       cytoskeleton   \n",
       "...                                                       ...   \n",
       "GO:0019867                                     outer membrane   \n",
       "GO:0022900                           electron transport chain   \n",
       "GO:0030031                           cell projection assembly   \n",
       "GO:0030234                          enzyme regulator activity   \n",
       "GO:0031012                               extracellular matrix   \n",
       "GO:0031982                                            vesicle   \n",
       "GO:0032196                                      transposition   \n",
       "GO:0034061                            DNA polymerase activity   \n",
       "GO:0034062                            RNA polymerase activity   \n",
       "GO:0042575                             DNA polymerase complex   \n",
       "GO:0042592                                homeostatic process   \n",
       "GO:0042597                                  periplasmic space   \n",
       "GO:0043190     ATP-binding cassette (ABC) transporter complex   \n",
       "GO:0043934                                        sporulation   \n",
       "GO:0044281                   small molecule metabolic process   \n",
       "GO:0044403  symbiosis, encompassing mutualism through para...   \n",
       "GO:0045182                     translation regulator activity   \n",
       "GO:0045333                               cellular respiration   \n",
       "GO:0046034                              ATP metabolic process   \n",
       "GO:0046039                              GTP metabolic process   \n",
       "GO:0046872                                  metal ion binding   \n",
       "GO:0048037                                   cofactor binding   \n",
       "GO:0048870                                      cell motility   \n",
       "GO:0051082                           unfolded protein binding   \n",
       "GO:0051186                         cofactor metabolic process   \n",
       "GO:0051301                                      cell division   \n",
       "GO:0055114                        oxidation-reduction process   \n",
       "GO:0070469                                  respiratory chain   \n",
       "GO:0071554               cell wall organization or biogenesis   \n",
       "GO:0071941                   nitrogen cycle metabolic process   \n",
       "\n",
       "            annotation counts  \n",
       "GO term                        \n",
       "GO:0000156                2.0  \n",
       "GO:0000160               50.0  \n",
       "GO:0000166              316.0  \n",
       "GO:0000746                3.0  \n",
       "GO:0000902                5.0  \n",
       "GO:0000988               17.0  \n",
       "GO:0001071               47.0  \n",
       "GO:0002376                0.0  \n",
       "GO:0003676              218.0  \n",
       "GO:0003774                2.0  \n",
       "GO:0003924                4.0  \n",
       "GO:0004386                6.0  \n",
       "GO:0004518               22.0  \n",
       "GO:0004672               18.0  \n",
       "GO:0004812               48.0  \n",
       "GO:0004872               78.0  \n",
       "GO:0005102                0.0  \n",
       "GO:0005215              150.0  \n",
       "GO:0005618                1.0  \n",
       "GO:0005634                0.0  \n",
       "GO:0005667                0.0  \n",
       "GO:0005694                2.0  \n",
       "GO:0005727                0.0  \n",
       "GO:0005737               83.0  \n",
       "GO:0005739                0.0  \n",
       "GO:0005773                0.0  \n",
       "GO:0005783                0.0  \n",
       "GO:0005794                0.0  \n",
       "GO:0005840               41.0  \n",
       "GO:0005856                0.0  \n",
       "...                       ...  \n",
       "GO:0019867               15.0  \n",
       "GO:0022900                8.0  \n",
       "GO:0030031                0.0  \n",
       "GO:0030234                3.0  \n",
       "GO:0031012                0.0  \n",
       "GO:0031982                0.0  \n",
       "GO:0032196                6.0  \n",
       "GO:0034061                3.0  \n",
       "GO:0034062               11.0  \n",
       "GO:0042575                1.0  \n",
       "GO:0042592               11.0  \n",
       "GO:0042597               16.0  \n",
       "GO:0043190                3.0  \n",
       "GO:0043934                0.0  \n",
       "GO:0044281              303.0  \n",
       "GO:0044403                0.0  \n",
       "GO:0045182                0.0  \n",
       "GO:0045333               19.0  \n",
       "GO:0046034                7.0  \n",
       "GO:0046039                1.0  \n",
       "GO:0046872              132.0  \n",
       "GO:0048037              183.0  \n",
       "GO:0048870                7.0  \n",
       "GO:0051082                1.0  \n",
       "GO:0051186               48.0  \n",
       "GO:0051301                5.0  \n",
       "GO:0055114              325.0  \n",
       "GO:0070469                0.0  \n",
       "GO:0071554                3.0  \n",
       "GO:0071941                4.0  \n",
       "\n",
       "[116 rows x 3 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = DataFrame(columns=('category', 'description', 'annotation counts'))\n",
    "df.index.name = 'GO term'\n",
    "\n",
    "with Session(API_BASE) as s:\n",
    "    f = Filter('page_size=100')\n",
    "    for ann in s.iterate('runs/SRR1047054/pipelines/2.0/go-slim', f):\n",
    "        df.loc[ann.accession] = [\n",
    "            ann.lineage, ann.description, ann.count\n",
    "        ]\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### List oceanic metagenomic samples collected in a temperature between 1°C and 5°C\n",
    "\n",
    "List samples: https://www.ebi.ac.uk/metagenomics/api/v0.2/biomes/root:Environmental:Aquatic:Marine:Oceanic/samples?experiment_type=metagenomic&metadata_key=temperature&metadata_value_gte=1&metadata_value_lte=5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample name</th>\n",
       "      <th>biome</th>\n",
       "      <th>temperature</th>\n",
       "      <th>depth</th>\n",
       "      <th>location</th>\n",
       "      <th>latitude</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRS981327</th>\n",
       "      <td>BP_381106-1</td>\n",
       "      <td>Marine</td>\n",
       "      <td>4.4889</td>\n",
       "      <td>None</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>28.7051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRS981330</th>\n",
       "      <td>BP_381105-1</td>\n",
       "      <td>Marine</td>\n",
       "      <td>5.5365</td>\n",
       "      <td>None</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>28.7051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRS981311</th>\n",
       "      <td>BP_381111-2</td>\n",
       "      <td>Marine</td>\n",
       "      <td>4.4199</td>\n",
       "      <td>None</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>28.6632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRS981308</th>\n",
       "      <td>BP_381112-2</td>\n",
       "      <td>Marine</td>\n",
       "      <td>4.281</td>\n",
       "      <td>None</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>28.6632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRS981307</th>\n",
       "      <td>BP_381112-3</td>\n",
       "      <td>Marine</td>\n",
       "      <td>4.281</td>\n",
       "      <td>None</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>28.6632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRS981326</th>\n",
       "      <td>BP_381106-2</td>\n",
       "      <td>Marine</td>\n",
       "      <td>4.4889</td>\n",
       "      <td>None</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>28.7051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRS981325</th>\n",
       "      <td>BP_381106-3</td>\n",
       "      <td>Marine</td>\n",
       "      <td>4.4889</td>\n",
       "      <td>None</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>28.7051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRS981309</th>\n",
       "      <td>BP_381112-1</td>\n",
       "      <td>Marine</td>\n",
       "      <td>4.281</td>\n",
       "      <td>None</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>28.6632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRS981322</th>\n",
       "      <td>BP_381107-3</td>\n",
       "      <td>Marine</td>\n",
       "      <td>4.2599</td>\n",
       "      <td>None</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>28.7051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRS981310</th>\n",
       "      <td>BP_381111-3</td>\n",
       "      <td>Marine</td>\n",
       "      <td>4.4199</td>\n",
       "      <td>None</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>28.6632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRS981328</th>\n",
       "      <td>BP_381105-3</td>\n",
       "      <td>Marine</td>\n",
       "      <td>5.5365</td>\n",
       "      <td>None</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>28.7051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRS981329</th>\n",
       "      <td>BP_381105-2</td>\n",
       "      <td>Marine</td>\n",
       "      <td>5.5365</td>\n",
       "      <td>None</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>28.7051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRS981312</th>\n",
       "      <td>BP_381111-1</td>\n",
       "      <td>Marine</td>\n",
       "      <td>4.4199</td>\n",
       "      <td>None</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>28.6632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRS981323</th>\n",
       "      <td>BP_381107-2</td>\n",
       "      <td>Marine</td>\n",
       "      <td>4.2599</td>\n",
       "      <td>None</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>28.7051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRS981324</th>\n",
       "      <td>BP_381107-1</td>\n",
       "      <td>Marine</td>\n",
       "      <td>4.2599</td>\n",
       "      <td>None</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>28.7051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERS1569005</th>\n",
       "      <td>Arctic seawater metagenome, 12 m</td>\n",
       "      <td>Oceanic</td>\n",
       "      <td>1.0376</td>\n",
       "      <td>12</td>\n",
       "      <td>Arctic Ocean</td>\n",
       "      <td>72.7978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERS1568998</th>\n",
       "      <td>Arctic seawater metagenome, 57 m</td>\n",
       "      <td>Oceanic</td>\n",
       "      <td>1.7079</td>\n",
       "      <td>57</td>\n",
       "      <td>Arctic Ocean</td>\n",
       "      <td>72.9989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERS1568997</th>\n",
       "      <td>Arctic seawater metagenome, 57 m</td>\n",
       "      <td>Oceanic</td>\n",
       "      <td>1.7079</td>\n",
       "      <td>57</td>\n",
       "      <td>Arctic Ocean</td>\n",
       "      <td>72.9989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERS1569006</th>\n",
       "      <td>Arctic seawater metagenome, 12 m</td>\n",
       "      <td>Oceanic</td>\n",
       "      <td>1.0376</td>\n",
       "      <td>12</td>\n",
       "      <td>Arctic Ocean</td>\n",
       "      <td>72.7978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERS1568999</th>\n",
       "      <td>Arctic seawater metagenome, 25 m</td>\n",
       "      <td>Oceanic</td>\n",
       "      <td>1.6423</td>\n",
       "      <td>25</td>\n",
       "      <td>Arctic Ocean</td>\n",
       "      <td>72.9989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERS493460</th>\n",
       "      <td>TARA_20111019T1928Z_133_Combined-EVENTS_CAST_M...</td>\n",
       "      <td>Oceanic</td>\n",
       "      <td>4.936634 °C</td>\n",
       "      <td>650 m</td>\n",
       "      <td>None</td>\n",
       "      <td>35.2698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERS494208</th>\n",
       "      <td>TARA_20120202T1839Z_145_Combined-EVENTS_CAST_M...</td>\n",
       "      <td>Oceanic</td>\n",
       "      <td>5.067239 °C</td>\n",
       "      <td>590 m</td>\n",
       "      <td>None</td>\n",
       "      <td>39.2392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERS240712</th>\n",
       "      <td>Landsort Depth 20090415 transect, sample ERS24...</td>\n",
       "      <td>Marine</td>\n",
       "      <td>4.1 °C</td>\n",
       "      <td>None</td>\n",
       "      <td>Baltic Sea</td>\n",
       "      <td>58.6000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERS240713</th>\n",
       "      <td>Landsort Depth 20090415 transect, sample ERS24...</td>\n",
       "      <td>Marine</td>\n",
       "      <td>5.23 °C</td>\n",
       "      <td>None</td>\n",
       "      <td>Baltic Sea</td>\n",
       "      <td>58.6000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERS240714</th>\n",
       "      <td>Landsort Depth 20090415 transect, sample ERS24...</td>\n",
       "      <td>Marine</td>\n",
       "      <td>5.72 °C</td>\n",
       "      <td>None</td>\n",
       "      <td>Baltic Sea</td>\n",
       "      <td>58.6000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRS004796</th>\n",
       "      <td>Brazos-Trinity Basin subseafloor sediment meta...</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>4 °C</td>\n",
       "      <td>1470 m</td>\n",
       "      <td>USA:Gulf of Mexico</td>\n",
       "      <td>27.3013</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  sample name     biome  \\\n",
       "accession                                                                 \n",
       "SRS981327                                         BP_381106-1    Marine   \n",
       "SRS981330                                         BP_381105-1    Marine   \n",
       "SRS981311                                         BP_381111-2    Marine   \n",
       "SRS981308                                         BP_381112-2    Marine   \n",
       "SRS981307                                         BP_381112-3    Marine   \n",
       "SRS981326                                         BP_381106-2    Marine   \n",
       "SRS981325                                         BP_381106-3    Marine   \n",
       "SRS981309                                         BP_381112-1    Marine   \n",
       "SRS981322                                         BP_381107-3    Marine   \n",
       "SRS981310                                         BP_381111-3    Marine   \n",
       "SRS981328                                         BP_381105-3    Marine   \n",
       "SRS981329                                         BP_381105-2    Marine   \n",
       "SRS981312                                         BP_381111-1    Marine   \n",
       "SRS981323                                         BP_381107-2    Marine   \n",
       "SRS981324                                         BP_381107-1    Marine   \n",
       "ERS1569005                   Arctic seawater metagenome, 12 m   Oceanic   \n",
       "ERS1568998                   Arctic seawater metagenome, 57 m   Oceanic   \n",
       "ERS1568997                   Arctic seawater metagenome, 57 m   Oceanic   \n",
       "ERS1569006                   Arctic seawater metagenome, 12 m   Oceanic   \n",
       "ERS1568999                   Arctic seawater metagenome, 25 m   Oceanic   \n",
       "ERS493460   TARA_20111019T1928Z_133_Combined-EVENTS_CAST_M...   Oceanic   \n",
       "ERS494208   TARA_20120202T1839Z_145_Combined-EVENTS_CAST_M...   Oceanic   \n",
       "ERS240712   Landsort Depth 20090415 transect, sample ERS24...    Marine   \n",
       "ERS240713   Landsort Depth 20090415 transect, sample ERS24...    Marine   \n",
       "ERS240714   Landsort Depth 20090415 transect, sample ERS24...    Marine   \n",
       "SRS004796   Brazos-Trinity Basin subseafloor sediment meta...  Sediment   \n",
       "\n",
       "            temperature   depth             location  latitude  \n",
       "accession                                                       \n",
       "SRS981327       4.4889     None  USA: Gulf of Mexico   28.7051  \n",
       "SRS981330       5.5365     None  USA: Gulf of Mexico   28.7051  \n",
       "SRS981311       4.4199     None  USA: Gulf of Mexico   28.6632  \n",
       "SRS981308        4.281     None  USA: Gulf of Mexico   28.6632  \n",
       "SRS981307        4.281     None  USA: Gulf of Mexico   28.6632  \n",
       "SRS981326       4.4889     None  USA: Gulf of Mexico   28.7051  \n",
       "SRS981325       4.4889     None  USA: Gulf of Mexico   28.7051  \n",
       "SRS981309        4.281     None  USA: Gulf of Mexico   28.6632  \n",
       "SRS981322       4.2599     None  USA: Gulf of Mexico   28.7051  \n",
       "SRS981310       4.4199     None  USA: Gulf of Mexico   28.6632  \n",
       "SRS981328       5.5365     None  USA: Gulf of Mexico   28.7051  \n",
       "SRS981329       5.5365     None  USA: Gulf of Mexico   28.7051  \n",
       "SRS981312       4.4199     None  USA: Gulf of Mexico   28.6632  \n",
       "SRS981323       4.2599     None  USA: Gulf of Mexico   28.7051  \n",
       "SRS981324       4.2599     None  USA: Gulf of Mexico   28.7051  \n",
       "ERS1569005      1.0376      12          Arctic Ocean   72.7978  \n",
       "ERS1568998      1.7079      57          Arctic Ocean   72.9989  \n",
       "ERS1568997      1.7079      57          Arctic Ocean   72.9989  \n",
       "ERS1569006      1.0376      12          Arctic Ocean   72.7978  \n",
       "ERS1568999      1.6423      25          Arctic Ocean   72.9989  \n",
       "ERS493460   4.936634 °C   650 m                 None   35.2698  \n",
       "ERS494208   5.067239 °C   590 m                 None   39.2392  \n",
       "ERS240712        4.1 °C    None           Baltic Sea   58.6000  \n",
       "ERS240713       5.23 °C    None           Baltic Sea   58.6000  \n",
       "ERS240714       5.72 °C    None           Baltic Sea   58.6000  \n",
       "SRS004796          4 °C  1470 m   USA:Gulf of Mexico   27.3013  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_metadata(metadata, key='temperature'):\n",
    "    import html\n",
    "    for m in metadata:\n",
    "        if m.var_name.lower() == key.lower():\n",
    "            value = m.var_value\n",
    "            unit = html.unescape(m.unit) if m.unit else \"\"\n",
    "            return \"{value} {unit}\".format(value=value, unit=unit)\n",
    "    return None\n",
    "\n",
    "depth_label = 'geographic location (depth)'\n",
    "temp_label = 'temperature'\n",
    "df = DataFrame(columns=('sample name', 'biome', 'temperature', 'depth', 'location', 'latitude'))\n",
    "df.index.name = 'accession'\n",
    "\n",
    "with Session(API_BASE) as s:\n",
    "    params = {\n",
    "        'experiment_type': 'metagenomic',\n",
    "        'metadata_key': 'temperature',\n",
    "        'metadata_value_gte': 1,\n",
    "        'metadata_value_lte': 5,\n",
    "        'latitude_gte': 0,\n",
    "        'include': 'biome,metadata',\n",
    "    }\n",
    "    f = Filter(urlencode(params))\n",
    "    for sample in s.iterate('biomes/root:Environmental:Aquatic:Marine/samples', f):\n",
    "        df.loc[sample.accession] = [\n",
    "            sample.sample_name, sample.biome.biome_name,\n",
    "            get_metadata(sample.metadata, temp_label),\n",
    "            get_metadata(sample.metadata, depth_label),\n",
    "            sample.geo_loc_name, sample.latitude\n",
    "        ]\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Export to CSV\n",
    "\n",
    "Get study: https://www.ebi.ac.uk/metagenomics/api/v0.2/studies/ERP005831"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>biome</th>\n",
       "      <th>lineage</th>\n",
       "      <th>longitude</th>\n",
       "      <th>latitude</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>study</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERP005831</th>\n",
       "      <td>ERS456668</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>root:Environmental:Aquatic:Freshwater:Lentic:S...</td>\n",
       "      <td>-1.56</td>\n",
       "      <td>52.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERP005831</th>\n",
       "      <td>ERS456669</td>\n",
       "      <td>Agricultural</td>\n",
       "      <td>root:Environmental:Terrestrial:Soil:Loam:Agric...</td>\n",
       "      <td>-1.61</td>\n",
       "      <td>52.19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              sample         biome  \\\n",
       "study                                \n",
       "ERP005831  ERS456668      Sediment   \n",
       "ERP005831  ERS456669  Agricultural   \n",
       "\n",
       "                                                     lineage  longitude  \\\n",
       "study                                                                     \n",
       "ERP005831  root:Environmental:Aquatic:Freshwater:Lentic:S...      -1.56   \n",
       "ERP005831  root:Environmental:Terrestrial:Soil:Loam:Agric...      -1.61   \n",
       "\n",
       "           latitude  \n",
       "study                \n",
       "ERP005831     52.38  \n",
       "ERP005831     52.19  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import csv\n",
    "\n",
    "with open(\"test_export.csv\", \"w\") as csvfile:\n",
    "    with Session(API_BASE) as s:\n",
    "        fieldnames = ['study', 'sample', 'biome', 'lineage', 'longitude', 'latitude']\n",
    "        writer = csv.DictWriter(csvfile, fieldnames=fieldnames)\n",
    "        writer.writeheader()\n",
    "        study = s.get('studies', 'ERP005831').resource\n",
    "        for sample in study.samples:\n",
    "            biome = sample.biome\n",
    "            row = {\n",
    "                'study': study.accession,\n",
    "                'sample': sample.accession,\n",
    "                'biome': biome.biome_name,\n",
    "                'lineage': biome.lineage,\n",
    "                'longitude': sample.longitude,\n",
    "                'latitude': sample.latitude\n",
    "            }\n",
    "            writer.writerow(row)\n",
    "\n",
    "df = DataFrame().from_csv('test_export.csv')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
